sweepbatcher/sweep_batch.go

```diff
@@ -460,16 +466,20 @@ func (b *batch) sweepExists(hash lntypes.Hash) bool {
 // Wait waits for the batch to gracefully stop.
 func (b *batch) Wait() {
 	b.log.Infof("Stopping")
-	b.wg.Wait()
```

Comment:
I doubt this has caused a race, wait groups are supposed to be used from different goroutines.See also implementation: https://cs.opensource.google/go/go/+/refs/tags/go1.22.4:src/sync/waitgroup.go;l=43

Reply:
If Add(1) and Wait() are called in parallel, the outcome for Wait() depends on the order of execution, which is a race condition: it can be either a no-op (if performed before Add), or a wait (if performed after). In practice, this means, that some goroutines of the batch could run away and outlive `batch.Wait()`.

See https://pkg.go.dev/sync#WaitGroup.Add

> If a WaitGroup is reused to wait for several independent sets of events, new Add calls must happen after all previous Wait calls have returned.

In the [implementation](https://cs.opensource.google/go/go/+/refs/tags/go1.22.4:src/sync/waitgroup.go;l=56-59) of WaitGroup.Add(1) there is the following part for race detection:

```go
	if race.Enabled && delta > 0 && v == int32(delta) {
		// The first increment must be synchronized with Wait.
		// Need to model this as a read, because there can be
		// several concurrent wg.counter transitions from 0.
		race.Read(unsafe.Pointer(&wg.sema))
	}
```

I think, this `race.Read` caused the crash above.

Reply:
Indeed we can't guarantee the order, so moving the `Wait()` into `Run`'s `defer` can be a good solution, however I think you'll also need to change `scheduleNextCall()` in order to correctly avoid adding a sweep while the batcher is shutting down.

Reply:
I added
```go
case <-b.finished:
  return func() {}, ErrBatchShuttingDown
```
to `scheduleNextCall()`, however it is not necessary in the current implementation, because channel `finished` is closed after channel `stopping` and there is already a case for `stopping`.

Also added TestSweepBatcherCloseDuringAdding to make sure adding a sweep during shutting down does not cause a crash. The test did not catch the originalrace condition.